File: python/data_structures/linkedlist/singly_linked_list_scratch.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class SinglyLinkedListScratch:

    def __init__(self):
        self.head = None
        self.tail = None
        self.size_value = 0

    def make_node(self, value):
        return Node(value)

    def push_front(self, value):
        n = self.make_node(value)
        n.next = self.head
        self.head = n

        if self.tail is None:
            self.tail = n

        self.size_value += 1
        return n

    def push_back(self, value):
        n = self.make_node(value)

        if self.tail is None:
            self.head = self.tail = n
        else:
            self.tail.next = n
            self.tail = n

        self.size_value += 1
        return n

    def insert_after(self, node, value):
        if node is None:
            raise ValueError("insert_after: node cannot be None")

        n = self.make_node(value)
        n.next = node.next
        node.next = n

        if self.tail is node:
            self.tail = n

        self.size_value += 1
        return n

    def pop_front(self):
        if self.head is None:
            raise IndexError("pop_front from empty list")

        value = self.head.value
        self.head = self.head.next
        self.size_value -= 1

        if self.head is None:
            self.tail = None

        return value

    def remove_after(self, node):
        if node is None or node.next is None:
            raise ValueError("remove_after: nothing to remove")

        to_remove = node.next
        node.next = to_remove.next

        if self.tail is to_remove:
            self.tail = node

        self.size_value -= 1
        return to_remove.value

    def find(self, value):
        current = self.head
        while current is not None:
            if current.value == value:
                return current
            current = current.next
        return None

    def is_empty(self):
        return self.size_value == 0

    def size(self):
        return self.size_value

File: java/data_structures/linkedlist/SinglyLinkedListScratch.java
package data_structures.linkedlist;

public class SinglyLinkedListScratch {

    public static class Node {
        public Object value;
        public Node next;

        public Node(Object value) {
            this.value = value;
            this.next = null;
        }
    }

    private Node head;
    private Node tail;
    private int sizeValue;

    public SinglyLinkedListScratch() {
        this.head = null;
        this.tail = null;
        this.sizeValue = 0;
    }

    private Node makeNode(Object value) {
        return new Node(value);
    }

    public Node pushFront(Object value) {
        Node n = makeNode(value);
        n.next = head;
        head = n;

        if (tail == null) {
            tail = n;
        }

        sizeValue++;
        return n;
    }

    public Node pushBack(Object value) {
        Node n = makeNode(value);

        if (tail == null) {
            head = tail = n;
        } else {
            tail.next = n;
            tail = n;
        }

        sizeValue++;
        return n;
    }

    public Node insertAfter(Node node, Object value) {
        if (node == null) {
            throw new IllegalArgumentException("insert_after: node cannot be null");
        }

        Node n = makeNode(value);
        n.next = node.next;
        node.next = n;

        if (tail == node) {
            tail = n;
        }

        sizeValue++;
        return n;
    }

    public Object popFront() {
        if (head == null) {
            throw new RuntimeException("pop_front from empty list");
        }

        Object value = head.value;
        head = head.next;
        sizeValue--;

        if (head == null) {
            tail = null;
        }

        return value;
    }

    public Object removeAfter(Node node) {
        if (node == null || node.next == null) {
            throw new IllegalArgumentException("remove_after: nothing to remove");
        }

        Node toRemove = node.next;
        node.next = toRemove.next;

        if (tail == toRemove) {
            tail = node;
        }

        sizeValue--;
        return toRemove.value;
    }

    public Node find(Object value) {
        Node current = head;
        while (current != null) {
            if ((current.value == null && value == null) ||
                (current.value != null && current.value.equals(value))) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    public boolean isEmpty() {
        return sizeValue == 0;
    }

    public int size() {
        return sizeValue;
    }

    public Node getHead() {
        return head;
    }

    public Node getTail() {
        return tail;
    }
}